coresimd/src/x86/x86_64/aes.rs

```diff
@@ -0,0 +1,144 @@
+pub unsafe fn _mm_aeskeygenassist_si128(a: __m128i, imm8: u8) -> __m128i {
+    macro_rules! call {
+        ($imm8:expr) => (aeskeygenassist(a, $imm8))
+    }
+    constify_imm8!(imm8, call)
+}
+
+#[cfg(test)]
+mod tests {
+    // The constants in the tests below are just bit patterns. They should not
+    // be interpreted as integers; signedness does not make sense for them, but
+    // __m128i happens to be defined in terms of signed integers.
+    #![allow(overflowing_literals)]
+
+    use stdsimd_test::simd_test;
+
+    use x86::*;
+
+    #[simd_test = "aes"]
+    unsafe fn test_mm_aesdec_si128() {
+        // Constants taken from https://msdn.microsoft.com/en-us/library/cc664949.aspx.
+        let a = __m128i(0x0123456789abcdef, 0x8899aabbccddeeff);
+        let k = __m128i(0x1133557799bbddff, 0x0022446688aaccee);
+        let e = __m128i(0x044e4f5176fec48f, 0xb57ecfa381da39ee);
+        let r = _mm_aesdec_si128(a, k);
+        assert_eq_m128i(r, e);
+    }
+
+    #[simd_test = "aes"]
+    unsafe fn test_mm_aesdeclast_si128() {
+        // Constants taken from https://msdn.microsoft.com/en-us/library/cc714178.aspx.
+        let a = __m128i(0x0123456789abcdef, 0x8899aabbccddeeff);
+        let k = __m128i(0x1133557799bbddff, 0x0022446688aaccee);
+        let e = __m128i(0x36cad57d9072bf9e, 0xf210dd981fa4a493);
+        let r = _mm_aesdeclast_si128(a, k);
+        assert_eq_m128i(r, e);
+    }
+
+    #[simd_test = "aes"]
+    unsafe fn test_mm_aesenc_si128() {
+        // Constants taken from https://msdn.microsoft.com/en-us/library/cc664810.aspx.
+        let a = __m128i(0x0123456789abcdef, 0x8899aabbccddeeff);
+        let k = __m128i(0x1133557799bbddff, 0x0022446688aaccee);
+        let e = __m128i(0x16ab0e57dfc442ed, 0x28e4ee1884504333);
+        let r = _mm_aesenc_si128(a, k);
+        assert_eq_m128i(r, e);
+    }
+
+    #[simd_test = "aes"]
+    unsafe fn test_mm_aesenclast_si128() {
+        // Constants taken from https://msdn.microsoft.com/en-us/library/cc714136.aspx.
+        let a = __m128i(0x0123456789abcdef, 0x8899aabbccddeeff);
+        let k = __m128i(0x1133557799bbddff, 0x0022446688aaccee);
+        let e = __m128i(0xb6dd7df25d7ab320, 0x4b04f98cf4c860f8);
+        let r = _mm_aesenclast_si128(a, k);
+        assert_eq_m128i(r, e);
+    }
+
+    #[simd_test = "aes"]
+    unsafe fn test_mm_aesimc_si128() {
+        // Constants taken from https://msdn.microsoft.com/en-us/library/cc714195.aspx.
+        let a = __m128i(0x0123456789abcdef, 0x8899aabbccddeeff);
+        let e = __m128i(0xc66c82284ee40aa0, 0x6633441122770055);
+        let r = _mm_aesimc_si128(a);
+        assert_eq_m128i(r, e);
+    }
+
+    #[simd_test = "aes"]
+    unsafe fn test_mm_aeskeygenassist_si128() {
+        // Constants taken from https://msdn.microsoft.com/en-us/library/cc714138.aspx.
+        let a = __m128i(0x0123456789abcdef, 0x8899aabbccddeeff);
+        let e = __m128i(0x857c266b7c266e85, 0xeac4eea9c4eeacea);
```

Comment:
Could this use `_mm_set_epi64x` to avoid making an assumption about the internal layout of `__m128i`?

Reply:
Done.